### PR TITLE
dev/mail#36 - Fix invalid unicode characters in bounce processing

### PR DESCRIPTION
**Pull request by tschuettler**
Overview
----------------------------------------
MySQL does not accept invalid unicode characters and therefore fails to write the bounce to the database if it is part of the bounce reason.
Issue tracker: https://lab.civicrm.org/dev/mail/issues/36

Before
----------------------------------------
Exception is thrown. Bounce is not stored in the datebase, but mail is moved to processed folder.

After
----------------------------------------
Invalid character are replaced with unicode replacement character. No exception. Bounce is saved.

Technical Details
----------------------------------------
Generating an invalid unicode character was not easy. Had to use a hex editor and replaced a valid character with an invalid one.

### CODE REVIEW

File: tests/phpunit/CRM/Utils/Mail/data/bounces/test_invalid_character.eml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+--==IFJRGLKFGIR18927894UHRUHIHD
+Content-Type: message/delivery-status
+
+Reporting-MTA: dns;srv.example.com
+
+Final-Recipient: rfc822;my@example.com
+Action: failed
+Status: 5.0.0
+Diagnostic-Code: X-Notes; Benutzer my (my@example.com) nicht im Domino-Verzeichnis aufgef�hrt
```
Comment (tschuettler) [inline tests/phpunit/CRM/Utils/Mail/data/bounces/test_invalid_character.eml]: The invalid character is a `\x81` in `aufgefhrt` between `f` and `h`.

**Comment by tschuettler** [comment]
Hmm, it looks like the unit tests inside that file never ran.

@eileenmcnaughton I guess this is due to class and file naming vs. the actual file path?

**Comment by eileenmcnaughton** [comment]
@tschuettler good spotting - change looks good too

**Comment by tschuettler** [comment]
@eileenmcnaughton Thanks for merging! :)

What do you think about replacing replacing non-utf8-mysql characters like emojis aswell? They will lead to the same error until we make the switch to utf8mb4.

I looked for other cases where the unit tests name did not match and created an issue for it:  https://lab.civicrm.org/dev/core/issues/647

**Comment by eileenmcnaughton** [comment]
@tschuettler I guess in this case (saving bounce reason) the data is relatively unimportant compared to the fact it's saved - ie. we are not putting this in outgoing messages but a technical field in the db - so if there is a chance other characters could break it it would be good to handle them

**Pull request merged**